### PR TITLE
Friendships v2: Implementing 2 rows for  mutual friendship

### PR DESCRIPTION
**Pull request by kirokoNjenga**
* Implementing 2 rows for mutual friendship 
* Added unit testing for friendship controller
**Testing every required function and made sure everything is working well**

### CODE REVIEW

**Comment by uryelah** [comment]
Hi **Samuel Njenga**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job so far but you are still missing the main requirement of this milestone:
**'when friendship is confirmed, an additional row with reversed user-friend pair is created'**

Read the checkpoint bellow carefully for details. You got this! 💪 

# Required Changes 🏗️ 🐛 💡
 
- [x] In your User#confirm_invites(user) method you should create a new Friendship role, this time with the user_id and friend_id inverted. The reason for that is that it will make it easier to get all the friends in the User#friends method, no need for concating at all.

-----

 **Kindly fix the pointed out issues before requesting the next code review**

------------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Comment by kirokoNjenga** [comment]
* already implemented the mutual friendship feature
image

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @Samkiroko, 👋 

### You still have not implemented the requirement of the last reviewer. 👇 

- [x] Firstly, don't merge your PR unless it has been approved, that way, we can see the exact change you made and easily point you in the right direction.

- [x] Also, what she meant was if I confirm you as a friend, I should be among `you.friends` and you should be among `me.friends`. This your code is one-sided, you can check it in the console to get a hang of what we mean.

**Kindly submit another code review request once this issue has been fixed.** _Happy coding!_ 🎉

**Comment by kirokoNjenga** [comment]
* Implement the requirement and is working on my App
*In case I missed something a sample code will be much appreciated
image

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes Required :muscle: 
Hi there @Samkiroko 
Great job on the project :clap: :clap: 
But we still have to address the confirm friend problem.

- [x] It seems that on last changes the invitation accept functionality stopped working. What the other TSEs were pointing out is that each friendship is saved as 2 rows on the Friendship table. When a user sends an invitation a row is created with the value false for the status column. When the other user accepts the invitation, you not only need to update the status column to true, but you also need to create a new row for the Friendship table. Please check out this [article](https://explainextended.com/2009/03/07/selecting-friends/) and this [post](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table) to read more about that.
Your code should look like something like this:
```ruby
def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

Please kindly submit for code review again after making this change.

Happy coding :computer: 
**Sergio Torres**

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Changes Required

Hi team :wave: 

I believe you are almost there. 

However, when I accept a friend request, I get an error.

Screenshot from 2020-06-16 08-33-02

Please use the above image as a reference, after fixing the issue, kindly submit another form for a code review.

#### ``` Have a great day ``` :slightly_smiling_face: 


``` Uchechukwu Henry Inyama ```

**Comment by kirokoNjenga** [comment]
* Cleared the error it working on my side
* In case of a different implementation working code will be appreciated.
```ruby
def confirm_invites(user)
    send_invitation&.update_attributes(confirmed: true)
    Friendship.create!(friend_id: send_invitation.user_id,
                  user_id: send_invitation.friend_id,
                  confirmed: true)
    friendship.save
  end
```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Nice Progress @Samkiroko :clap: However, some changes need to be applied in order to get approval for this milestone. Check :arrow_down: 

image

- App crashes when accepting a friendship
- I noticed that you treated friendship variable as local and instance variable at the same time. Instance variables are only recommended in controllers, but in models, conventionally we use local variables if needed.

- Use [Scopes](https://www.rubyguides.com/2019/10/scopes-in-ruby-on-rails/), ActiveRecord methods & queries to find a particular record on the DB.

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by kirokoNjenga** [comment]
* solved the issue 
* Again I have taken so much time on this issue please if you have a different solution than mine, please provide a working code am sure I can learn from that, thank you.
** I need to move on **

File: app/models/user.rb [hunk 1]
```diff
@@ -42,10 +42,8 @@ def send_invitation(user_id)
   end
 
   def confirm_invites(user)
-    send_invitation.update_attributes(confirmed: true)
-    Friendship.create!(friend_id: send_invitation.user_id,
-                  user_id: send_invitation.friend_id,
-                  confirmed: true)
```
Comment (nikasulo) [inline app/models/user.rb]: What you had here was 'better' in a sense. What was missing is the fact that you tried to run update attributes on a method. Instead, you should try to find the inverse friendship corresponding to the current user and the user in question and update that. Then create an inverse friendship like you attempted to in 45-48

Comment (nikasulo) [reply app/models/user.rb]:
```ruby
def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

Comment (nikasulo) [reply app/models/user.rb]: This should be done on the friendship

**Review by nikasulo — CHANGES_REQUESTED** [review]
# Nice work so far @Samkiroko 
Please consider my comments below, make the required change and request another review, I wish you the very best and see you at the top 🚀

File: app/models/user.rb [hunk 2]
```diff
@@ -42,10 +42,8 @@ def send_invitation(user_id)
...
+    friendship = inverse_friendships.where(user_id: user).first
+    friendship.status = true
```
Comment (nikasulo) [inline app/models/user.rb]: Please address this. The reason why it's a bit problematic is you are trying to confirm the friendship in the user model when the logic should be happening on the friendship itself

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required 
### Great effort @Samkiroko  so far :clap: 
We will take a detailed look on what is required of us in this milestone.

**REQUIRED**
* When friendship is confirmed, an additional row with reversed user-friend pair is created.
Currently, when I accept a friendship record no additional row has been created. :point_down: 
Screenshot from 2020-06-17 21-43-11
After taking some time to look at our code, I realized that the validation we have in the Friendship model is the one that is preventing us to create an inverse row with a confirmed value set to true.
Let's get rid of 
  ```ruby
    validate :friendship_valid?
  ```
  code in our friendship model.
* Having two rows for mutual friendship will enable us to  find users friends only using `user_id` in the friendship table.
  ```ruby
    has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
    has_many :friends, through: :confirmed_friendships
  ```
  We can now get rid on `friends` methods that used ruby objects instead of queries to find users friends.         Remember the database queries was designed to be more efficient than Ruby for this purpose.
* We can also modify `pending_friends` User model method to make use of queries instead of passing objects to the application memory.
  ```ruby
    has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
    has_many :pending_friends, through: :pending_friendships, source: :friend
  ```
* Lastly, we can update `friend_requests` method to make use of database queries. We can find friend requests from user_id in `inverse_friendships` (where current_user is referenced as friend) with a confirmed status of `false`, we can write the query as follows:
  ```ruby
    has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
    has_many :friend_requests, through: :inverted_friendships, source: :user
  ```

That's all we need to make this feature more efficient.

**After working on the changes, submit another request for review.**

_Feel free to react out via slack for any issue/clarification._

**Happy Coding :computer: 
Simon Wathigo**

**Comment by kirokoNjenga** [comment]
* deed what I was advised if the is a different opinion share your code, I think I have spent more day on this single issue than the project itself.  
* Everyone has their own way 
* hope I can move on now instead make the feature more efficient

File: app/models/friendship.rb [hunk 3]
```diff
@@ -1,4 +1,25 @@
 class Friendship < ApplicationRecord
   belongs_to :user
   belongs_to :friend, class_name: 'User'
+
+  #validate :friendship_valid?
```
Comment (kobiyoyo) [inline app/models/friendship.rb:5]: Lets remove this

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello Team,  You have done a great job with this project. :clap:

#### You can go ahead and merge !!!

**Pull request merged**